File: packages/packagechk/src/packagechk/check.py
from __future__ import annotations

from packagechk.gateways.registries.gateway import PackageRegistryGateway
from packagechk.models import PackageCheckReport, Registry, RegistryCheckResult

DEFAULT_REGISTRIES: tuple[Registry, ...] = (Registry.PYPI, Registry.NPM)


def registry_selection(registry_options: tuple[str, ...]) -> tuple[Registry, ...]:
    if not registry_options:
        return DEFAULT_REGISTRIES
    return tuple(Registry(option) for option in registry_options)


def check_package_name(
    *,
    package_name: str,
    registries: tuple[Registry, ...],
    registry_gateway: PackageRegistryGateway,
) -> PackageCheckReport:
    results: list[RegistryCheckResult] = []
    for registry in registries:
        if registry is Registry.PYPI:
            results.append(registry_gateway.check_pypi(package_name))
        elif registry is Registry.NPM:
            results.append(registry_gateway.check_npm(package_name))
        elif registry is Registry.BREW:
            results.append(registry_gateway.check_brew(package_name))
        else:
            results.append(
                RegistryCheckResult.unsupported(
                    registry,
                    input_name=package_name,
                    message=f"{registry.value} availability checks are not supported",
                )
            )
    return PackageCheckReport(input_name=package_name, results=tuple(results))

File: ts/packages/packagechk/src/check.ts
import {
	REGISTRIES,
	type PackageCheckReport,
	type Registry,
	type RegistryCheckResult,
} from "./models.ts";
import type { PackageRegistryGateway } from "./registry-gateways.ts";

export function registrySelection(registryOptions: readonly Registry[]): readonly Registry[] {
	return registryOptions.length === 0 ? REGISTRIES : [...registryOptions];
}

export async function checkPackageName(options: {
	packageName: string;
	registries: readonly Registry[];
	registryGateway: PackageRegistryGateway;
}): Promise<PackageCheckReport> {
	const results = await Promise.all(
		options.registries.map((registry) =>
			checkRegistry({
				packageName: options.packageName,
				registry,
				registryGateway: options.registryGateway,
			}),
		),
	);
	return { inputName: options.packageName, results };
}

async function checkRegistry(options: {
	packageName: string;
	registry: Registry;
	registryGateway: PackageRegistryGateway;
}): Promise<RegistryCheckResult> {
	switch (options.registry) {
		case "pypi":
			return await options.registryGateway.checkPypi(options.packageName);
		case "npm":
			return await options.registryGateway.checkNpm(options.packageName);
		case "brew":
			return await options.registryGateway.checkBrew(options.packageName);
	}
	const exhaustive: never = options.registry;
	return exhaustive;
}
